src/visitors/ast.rs

```diff
@@ -121,3 +121,188 @@ impl From<&Package> for PackageInfo {
+pub(crate) const SEMANTIC_TOKEN_NUMBER: u32 = 1;
+pub(crate) const SEMANTIC_TOKEN_STRING: u32 = 2;
+
+#[derive(Clone, Default)]
+pub struct SemanticTokenVisitor {
+    pub tokens: Vec<lsp::SemanticToken>,
+}
+
+impl<'a> Visitor<'a> for SemanticTokenVisitor {
+    fn visit(&mut self, node: walk::Node<'a>) -> bool {
+        match node {
+            walk::Node::PackageClause(node) => {
+                let location = node.base.location.clone();
+                let delta_line = location.start.line;
+                let delta_start = location.start.column;
+                let length = 7; // Length of "package"
+                self.tokens.push(lsp::SemanticToken {
+                    delta_line,
+                    delta_start,
+                    length,
+                    token_type: SEMANTIC_TOKEN_KEYWORD,
+                    token_modifiers_bitset: 0,
+                });
+
+                // Identifier itself is just too broad. Manually handle
+                // the case for the string identifier.
+                let name_location = node.name.base.location.clone();
+                let name_delta_line = name_location.start.line;
+                let name_delta_start = name_location.start.column;
+                let length = name_location.end.column
+                    - name_location.start.column;
+                self.tokens.push(lsp::SemanticToken {
+                    delta_line: name_delta_line,
+                    delta_start: name_delta_start,
+                    length,
+                    token_type: SEMANTIC_TOKEN_STRING,
+                    token_modifiers_bitset: 0,
+                });
+            }
+            walk::Node::ImportDeclaration(declaration) => {
+                let location = declaration.base.location.clone();
+                let delta_line = location.start.line;
+                let delta_start = location.start.column;
+                let length = 6; // Length of "import"
+                self.tokens.push(lsp::SemanticToken {
+                    delta_line,
+                    delta_start,
+                    length,
+                    token_type: SEMANTIC_TOKEN_KEYWORD,
+                    token_modifiers_bitset: 0,
+                });
+            }
+            walk::Node::IntegerLit(literal) => {
+                let location = literal.base.location.clone();
+                let delta_line = location.start.line;
+                let delta_start = location.start.column;
+                let length =
+                    location.end.column - location.start.column;
+                self.tokens.push(lsp::SemanticToken {
+                    delta_line,
+                    delta_start,
+                    length,
+                    token_type: SEMANTIC_TOKEN_NUMBER,
+                    token_modifiers_bitset: 0,
+                });
+            }
+            walk::Node::FloatLit(literal) => {
+                let location = literal.base.location.clone();
```

Comment:
nit: another potential needless clone here

:microphone: I'll be here all night...

Looks like there's one or more of these in each match arm.